server/proxy/sockproxy/sockproxy.go

```diff
@@ -0,0 +1,229 @@
+		c.lg.Error("error reading from the connection", zap.Error(err))
+		conn.Close()
+		return
+	}
+	// is TLS
+	if b[0] == 0x16 {
+		if !c.secure {
+			c.lg.Error("secure connections not enabled")
+			conn.Close()
+			return
+		}
+		// Establish the TLS connection
+		tlsConn := tls.Server(buffConn, c.tlsConfig)
+		err = tlsConn.Handshake()
+		if err != nil {
+			c.lg.Error("error establishing the TLS connection", zap.Error(err))
+			conn.Close()
+			return
+		}
+
+		proxiedConn = newBufferedConn(tlsConn)
+		// TODO: verify this heuristic, use 24 bytes to be safe.
+		// 33 is the magic number to get the  [FrameHeader SETTINGS len=0] sent by GRPCs connection
+		// It is a "new" connection obtained after the handshake so we have to do another read
+		// to get the new data, but be careful to not get blocked in the read if there are no enough data.
+		b, err = proxiedConn.Peek(24)
+		if err != nil {
+			// try to see how far we go, it will be discarded by the server if we route it to the wrong one
+			c.lg.Error("error reading from the TLS connection", zap.Error(err))
+		}
+	} else {
+		if !c.insecure {
+			c.lg.Info("insecure connections not enabled")
+			conn.Close()
+			return
+		}
+		proxiedConn = buffConn
+		// read all the remaining buffer, we already read 1 byte
+		b, err = proxiedConn.Peek(buffConn.r.Buffered())
+		if err != nil {
+			// try to see how far we go, it will be discarded by the server if we route it to the wrong one
+			c.lg.Error("error reading from the connection", zap.Error(err))
+		}
+
+	}
+	// Checking if is HTTP, since we don't support HTTP/2 with Prior Knowledge
+	// the first connection has to be HTTP/1.1 with an Upgrade request.
+	if bytes.Contains(b, []byte("HTTP/1.1")) && c.http != nil {
+		c.http.connc <- proxiedConn
+		return
+	}
+
+	// if we didn't detect http send it to the grpc server if it is configured
+	if c.grpc != nil {
+		c.grpc.connc <- proxiedConn
+		return
+	}
+
+	log.Println("unknown connection")
+	conn.Close()
+}
+
+// Close closes the listeners
+func (c *SockProxy) Close() error {
+	c.closeDoneChans()
+	return c.root.Close()
+}
+
+func (c *SockProxy) closeDoneChans() {
+	select {
+	case <-c.donec:
+	default:
+		close(c.donec)
+	}
+}
+
+// muxListener is the listener exposed to the HTTP and GRPC servers
+// The root listener Accept() method is overriden
+// The multiplexed servers have access to the Close() and Address()
+// methods on the root listener.
+type muxListener struct {
+	net.Listener
+	connc chan net.Conn
+	donec chan struct{}
+}
+
+var _ net.Listener = (*muxListener)(nil)
+
+func (l muxListener) Accept() (net.Conn, error) {
+	select {
+	case c, ok := <-l.connc:
+		if !ok {
+			return nil, net.ErrClosed
+		}
+		return c, nil
+	case <-l.donec:
+		return nil, net.ErrClosed
+	}
+}
+
+// NewHTTPListener returns a HTTP listener to forward http requests
+func (c *SockProxy) HTTPListener() net.Listener {
+	if c.http == nil {
+		c.http = &muxListener{c.root, make(chan net.Conn), c.donec}
```

Comment:
setting `c.http` would race against `c.Serve`. Maybe we could:
* Document/forbid calling HTTPListener after `c.Serve` was called
* Sync via mutex.

Same for GRPCListener